lib/internal/modules/esm/loader.js

```diff
@@ -351,12 +351,21 @@ class ESMLoader {
       // The request & response have already settled, so they are in
       // fetchModule's cache, in which case, fetchModule returns
       // immediately and synchronously
-      url = fetchModule(new URL(url), { parentURL: url }).resolvedHREF;
-      // This should only occur if the module hasn't been fetched yet
-      if (typeof url !== 'string') { // [2]
-        throw new ERR_INTERNAL_ASSERTION(`Base url for module ${url} not loaded.`);
+      const module = fetchModule(new URL(url), { parentURL: url });
+
+      if (typeof module?.resolvedHREF === 'string') { // [2]
+        url = module.resolvedHREF;
+      } else { // This should only occur if the module hasn't been fetched yet
+        if (getOptionValue('--experimental-network-imports')) {
+          throw new ERR_INTERNAL_ASSERTION(
+            `Base url for module ${url} not loaded.`
+          );
+        } else {
+          url = module;
```

Comment:
Is `url` supposed to be a string or a `URL` or something else? Can we avoid the polymorphism?

Reply:
I think `module` here _is_ a url string (that's why it failed `if (typeof module?.resolvedHREF === 'string')` above). I'll confirm when I write the test cases.

Reply:
It feels like this bug happened in the first place because it wasn’t very clear in the code what this object was supposed to be (?) so maybe we could pick better names and/or add JSDoc types to make it much more obvious?

Reply:
Okay, I lied. `module` here is the unsettled promise, which needs to be surfaced to the caller, which _can_ await the promise (but `getBaseURL()` cannot).

I don't know if this polymorphism is avoidable because of the async vs sync constraints. If it is possible, I think this might require a very large refactor of `network-imports`.

Reply:
Firstly, thanks so much @JakobJingleheimer for coming through on this, it is really very much appreciated. Great to see it is a relatively straightforward fix.

It would be possible to avoid the asynchronous architecture here in `getBaseURL` by just updating the original module jobs to have two separate properties - a `url` and `responseUrl` property as two separate fields when the original parent module job is created. Then for a given module job, either can be picked synchronously depending on the situation (`import.meta.url` v parentUrl resolution). Then there is no need for a `getBaseURL` architecture or any asynchronousity there, which will also be necessary when it comes to supporting a sync resolver.

If the above is a larger refactoring that you don't have time for @JakobJingleheimer I would suggest we land this fix at least to resolve the bug, since the architecture question simply is the nature of the current implementation.

Reply:
@guybedford’s recommendation sounds good to me. If you have the time, perhaps a few JSDoc comments to the existing code could be slipped into this PR? So that when we get to the refactor it can hopefully go a little smoother.

Reply:
I added a test case (and confirmed it fails on `master`) to cover this bug (and it passes in this PR branch). That ended up taking my whole evening/night (`spawnSync()` was pulling the rug out from under me), so I didn't get a chance to try Guy's suggestion. I'll try to get to it today; if not, I won't get to it til next weekend.

p.s. thank you @Trott for saving my sanity.

Reply:
I've patched the contract here to rather treat any custom loader not using internal import fetches as simply not supporting a separate response URL (since they can't since they have no hooks). As a result the resultant URL is just the identity and we have well-defined function signatures all around.